File: definitive/02_GentleIntroduction/FlightExample.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.max


object FlightExample {
  val DATA = "../../data"

  spark = SparkSession.builder.appName("flights").getOrCreate()
  val data = spark
    .read
    .option("inferScehma", "true")
    .option("header", "true")
    .csv(s"$DATA/flight-data/csv/2015-summary.csv")

  // DataFrames and SQL
  data.createOrReplaceTempView("flightData")

  val sqlWay = spark.sql("""
    SELECT DEST_COUNTRY_NAME, COUNT(1)
    FROM flightData
    GROUP BY DEST_COUNTRY_NAME""")
  val dfWay = data.groupBy('DEST_COUNTRY_NAME).count()
  println(sqlWay.explain)
  println(dfWay.explain)

  println(spark.sql("SELECT MAX(count) FROM flightData").take(1))
  println(data.select(max("count")).take(1))

  val maxSql = spark.sql("""
    SELECT DEST_COUNTRY_NAME, SUM(count) AS dest_total
    FROM flightData
    GROUP BY DEST_COUNTRY_NAME
    ORDER BY SUM(count) DESC
    LIMIT 5""")
  maxSql.show()
}

File: definitive/02_GentleIntroduction/flight_example.py
from pyspark.sql import SparkSession


DATA = '../../data'

spark = SparkSession.builder.appName('flights').getOrCreate()
data = (spark
        .read
        .option('inferScehma', 'true')
        .option('header', 'true')
        .csv(f'{DATA}/flight-data/csv/2015-summary.csv'))
print(data.take(3))
data.sort('count').explain()

spark.conf.set('spark.sql.shuffle.partitions', '5') # default is 200
print(data.sort('count').take(2))


# DataFrames and SQL
data.createOrReplaceTempView('flight_data')
sql_way = spark.sql('''
    SELECT DEST_COUNTRY_NAME, COUNT(1)
    FROM flight_data
    GROUP BY DEST_COUNTRY_NAME''')

df_way = data.groupBy('DEST_COUNTRY_NAME').count()

print('sql:\n')
sql_way.explain()
print('\ndf:\n')
df_way.explain()




